src/types/typeobject.rs

```diff
@@ -11,38 +13,141 @@ pub struct PyType(PyAny);
 pyobject_native_type_core!(PyType, pyobject_native_static_type_object!(ffi::PyType_Type), #checkfunction=ffi::PyType_Check);
 
 impl PyType {
-    /// Creates a new type object.
+    /// Deprecated form of [`PyType::new_bound`].
     #[inline]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`new` will be replaced by `new_bound` in a future PyO3 version"
+        )
+    )]
     pub fn new<T: PyTypeInfo>(py: Python<'_>) -> &PyType {
         T::type_object_bound(py).into_gil_ref()
     }
 
+    /// Creates a new type object.
+    #[inline]
+    pub fn new_bound<T: PyTypeInfo>(py: Python<'_>) -> Bound<'_, PyType> {
+        T::type_object_bound(py)
+    }
+
     /// Retrieves the underlying FFI pointer associated with this Python object.
     #[inline]
     pub fn as_type_ptr(&self) -> *mut ffi::PyTypeObject {
-        self.as_ptr() as *mut ffi::PyTypeObject
+        self.as_borrowed().as_type_ptr()
     }
 
-    /// Retrieves the `PyType` instance for the given FFI pointer.
+    /// Deprecated form of [`PyType::from_type_ptr_bound`].
     ///
     /// # Safety
-    /// - The pointer must be non-null.
-    /// - The pointer must be valid for the entire of the lifetime for which the reference is used.
+    ///
+    /// - The pointer must a valid non-null reference to a `PyTypeObject`.
     #[inline]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`from_type_ptr` will be replaced by `from_type_ptr_bound` in a future PyO3 version"
+        )
+    )]
     pub unsafe fn from_type_ptr(py: Python<'_>, p: *mut ffi::PyTypeObject) -> &PyType {
-        py.from_borrowed_ptr(p as *mut ffi::PyObject)
+        Self::from_type_ptr_bound(py, &p).to_owned().into_gil_ref()
+    }
+
+    /// Converts the given FFI pointer into `&Bound<PyType>`, to use in safe code.
+    ///
+    /// # Safety
+    /// - The pointer must be non-null.
+    /// - The pointer must be valid for the entire of the lifetime 'a for which the reference is used,
+    ///   as with `std::slice::from_raw_parts`.
+    #[inline]
+    pub unsafe fn from_type_ptr_bound<'a, 'py>(
+        py: Python<'py>,
+        p: &'a *mut ffi::PyTypeObject,
+    ) -> &'a Bound<'py, PyType> {
+        let object_ptr = &*(p as *const *mut ffi::PyTypeObject as *const *mut ffi::PyObject);
+        Bound::ref_from_ptr(py, object_ptr).downcast_unchecked()
```

Comment:
Is this usage common enough, that we don't simply want to take ownership of the pointer? `& *mut _` just looks a bit odd to me, but I'm fine with it, if we want to provide a borrowing API.

Reply:
At one point this was used in the macro implementation, but that has now been removed with the changes in #3831. This is now only used in `PyAnyMethods::get_type`, which immediately makes it owned.

So I think this is a fair concern, let's just make this owned and keep things simple until users show us that borrowing here is performance critical (and their borrow is safe).